help_function/src/util_group_store_test.cpp

```diff
@@ -0,0 +1,304 @@
+#include <sycl/sycl.hpp>
+
+
+template <dpct::group::store_algorithm S>
+bool helper_validation_function(const int *ptr, const char *func_name) {
+  if constexpr (S == dpct::group::store_algorithm::BLOCK_STORE_DIRECT) {
+    for (int i = 0; i < 512; ++i) {
+      if (ptr[i] != i) {
+        std::cout << func_name << "_blocked"
+                  << " failed\n";
+        std::ostream_iterator<int> Iter(std::cout, ", ");
+        std::copy(ptr, ptr + 512, Iter);
+        std::cout << std::endl;
+        return false;
+      }
+    }
+    std::cout << func_name << "_blocked"
+              << " pass\n";
+  } else {
+    int expected[512];
+    int num_threads = 128;
+    int items_per_thread = 4;
+    for (int i = 0; i < num_threads; ++i) {
+      for (int j = 0; j < items_per_thread; ++j) {
+        expected[i * items_per_thread + j] = j * num_threads + i;
+      }
+    }
+    for (int i = 0; i < 512; ++i) {
+      if (ptr[i] != expected[i]) {
+        std::cout << func_name << "_striped"
+                  << " failed\n";
+        std::ostream_iterator<int> Iter(std::cout, ", ");
+        std::copy(ptr, ptr + 512, Iter);
+        std::cout << std::endl;
+        return false;
+      }
+    }
+    std::cout << func_name << "_striped"
+              << " pass\n";
+  }
+  return true;
+}
+
+bool subgroup_helper_validation_function(const int *ptr, const uint32_t *sg_sz,
+                                         const char *func_name) {
+  
+  int expected[512];
+  int num_threads = 128;
+  int items_per_thread = 4;
+  uint32_t sg_sz_val = *sg_sz;
+  for (int i = 0; i < num_threads; ++i) {
+    for (int j = 0; j < items_per_thread; ++j) {
+      expected[items_per_thread * i + j] =
+          (i / sg_sz_val) * sg_sz_val * items_per_thread + sg_sz_val * j +
+          i % sg_sz_val;
+    }
+  }
+  
+  for (int i = 0; i < 512; ++i) {
+    if (ptr[i] != i) {
+      std::cout << " failed\n";
+      std::ostream_iterator<int> Iter(std::cout, ", ");
+      std::copy(ptr, ptr + 512, Iter);
+      std::cout << std::endl;
+      return false;
+    }
+  }
+
+  std::cout << func_name << " pass\n";
+  return true;
+}
+
+template < dpct::group::store_algorithm S> bool test_group_store() {
+  // Tests dpct::group::load_algorithm::BLOCK_LOAD_DIRECT 
+  // dpct::group::load_algorithm::BLOCK_LOAD_STRIPED in its entirety as API
+  // functions
+  // Tests dpct::group::store_algorithm::BLOCK_STORE_DIRECT 
+  // dpct::group::store_algorithm::BLOCK_STORE_STRIPED in its entirety as API
+  // functions
+  sycl::queue q(dpct::get_default_queue());
+  int data[512];
+  int num_threads = 128;
+  int items_per_thread = 4;
+  if constexpr(S == dpct::group::store_algorithm::BLOCK_STORE_DIRECT) {
+    for (int i = 0; i < 512; i++) {data[i] = i;}
+  }
+  else{
+    for (int i = 0; i < num_threads; ++i) {
+      for (int j = 0; j < items_per_thread; ++j) {
+        data[i * items_per_thread + j] = j * num_threads + i;        
+      }
+      
+    }
+  }

```

Comment:
Does this pass?  It seems like you are "encoding" the store index map twice here in the test for striped store patterns.  
1) in the data initialization
2) in the verification

It seems like this should not pass because you basically have the following:

- init: shuffled data
- load: direct (still shuffled data)
- store: striped (unshuffle the data to be in order)
- verify: unshuffled (in order) data verified against shuffled data

For direct, it makes sense because the data stays unshuffled the whole time.  

From my perspective it seems that either the data should always be initialized in order, or always verified in order regardless of the store alg.

Reply:
Yes this one I also think arranging the initializer in order makes sense,
The idea was to mimic the effect of load operation through the manual arrangements of the input buffer.
All cases pass expect a standalone striped store which I am debugging.

Reply:
"The idea was to mimic the effect of load operation through the manual arrangements of the input buffer."
This makes sense, but if you do it this way, then the verification would need to be in-order (similar to the other PR).  

If this is passing, then either I'm missing something or there is some bug.